### DISCUSSION TITLE
Использование URN для Identifier.system и Coding.system

### DISCUSSION THREAD

**Discussion by niquola** [fhir-ru/core#61] [General]
Предложение использовать [URN](https://ru.m.wikipedia.org/wiki/URN)  -   Uniform Resource Name -  для кодирования system в идентификаторах организаций и пациентов.

> URN указывает неизменное имя ресурса без указания его местонахождения и способа обращения. В результате URN-имена постоянны, они не зависят от конкретных серверов и протоколов. Другими словами, URN концептуально обозначает сам ресурс, а не место

URN будут короче и позволят не привязываться к конкретному адресу и протоколу.

Пример:

* `urn:ru:id:ogrn` - ОГРН  организации 
* `urn:ru:id:snils` - СНИЛС пациента

Альтернативным решением является использование URL.

Плюсы/минусы URL - с точностью до наоборот -  он длинее (нужно думать про протокол), но его можно открыть в браузере

**Comment by niquola** [comment]
просто пальчик не катит - нужно полный комментарий/возражение

**Reply by vadi2** [reply]
на мобиле был, но https://github.com/fhir-ru/core/discussions/61#discussioncomment-1810796 подходит!

Мои 2с что чуть больше работы этого стоит и себя отплачивает.

**Comment by Vrodvi** [comment]
С представленным обозначением urn'ов не согласен.

Конкретно для СНИЛС и ОГРН, например, уже есть OID'ы и можно использовать urn с OID'ом (обсуждали в https://github.com/fhir-ru/core/discussions/41), т.е. через префикс `urn:oid:`.

Также, нужно посмотреть на имеющийся реестр OID'ов - https://nsi.rosminzdrav.ru/#!/oidRepository (аналог https://www.hl7.org/oid/index.cfm), в нём можно вести все необходимые объекты и давать на них ссылки в виде URL'ов, к примеру для идентификаторов организаций по ФРМО 
`https://nsi.rosminzdrav.ru/#!/oidRepository/1/1.2.643.5.1.13.13.11.1461`

| identifier | URL | URI | OID (for non-FHIR systems) | Type | Comment
| --- | --- | --- | --- | --- | --- |
| Глобально уникальный идентификатор медицинской организации по Реестру медицинских организаций Российской Федерации | https://nsi.rosminzdrav.ru/#!/oidRepository/1/1.2.643.5.1.13.13.11.1461 | urn:oid:1.2.643.5.1.13.13.11.1461 | 1.2.643.5.1.13.13.11.1461 |   |  
| Идентификационный номер налогоплательщика (ИНН ) | https://nsi.rosminzdrav.ru/#!/oidRepository/1/1.2.643.3.131.1.1 ??? | urn:oid:1.2.643.3.131.1.1 | 1.2.643.3.131.1.1 |   |
| Основной государственный регистрационный номер (ОГРН) | https://nsi.rosminzdrav.ru/#!/oidRepository/1/1.2.643.100.1 ??? | urn:oid:1.2.643.100.1 | 1.2.643.100.1 |   |
| СНИЛС | https://nsi.rosminzdrav.ru/#!/oidRepository/1/1.2.643.100.3 ??? | urn:oid:1.2.643.100.3 | 1.2.643.100.3 |

У варианта с URL преимущество в виде возможности перейти по адресу и получить дополнительные полезные метаданные, особенно если будет вестись полноценный реестр.

Также с вариантом через OID проще вести сам идентификатор и не закладывать смысл в его структуру - структура OID не значащая, а весь смысл можно описать в карточке этого OID в реестре.
Также при заведении OID понятен принцип ответственности - известно кто имеет право создавать дочерние узлы в OID, в варианте с URN это нужно будет оговаривать, чтобы избежать пересечения.
 
Кроме того в варианте с urn осмысленная структура может быть и отрицательным качеством, т.к. потребует больших усилий на согласование регламента структуры этих urn, так, например:
почему предложено именно `urn:ru:id:`, а не 
`urn:fhir-ru:id:`
`urn:ru:fhir:id:`
`urn:ru:fhir-ru:id:`
`urn:ru:fhir-ru:core:id:`
или аналогичные варианты?

**Reply by niquola** [reply]
Проблема с oid, как раз в их нечитабельности - это постоянная мука в v2 и x12 - каждый раз расшифровывать чиселки

**Comment by ai-fhir** [comment]
Продублирую цитату из [спецификации](http://www.hl7.org/fhir/datatypes.html#Identifier): 

> Resolvable URLs are generally preferred by implementers over non-resolvable URNs, particularly opaque URNs such as OIDs (urn:oid:) or UUIDs (urn:uuid:).

Плюсы URL:
1. Они тоже могут быть понятными без дополнительных действий, если URL будут человекочитаемые (url, в котором закодирован только oid, сложно идентифицировать при беглом просмотре)
2. URL **должен** вести на страницу описания кодовой системы, что облегчает поиск дополнительной информации

Минусы URL:
1. Более громоздкие чем может быть urn
2. Схема - http/https - не несет смысла и в ней легко сделать ошибку, а поддерживать оба варианта - добавляет сложность принимающей системе

Мой приоритет такой:
1. Человекочитаемый URL (т.к. это предпочтение стандарта)
2. Человекочитаемый URN (с небольшим отставанием)
3. Остальные варианты

**Reply by niquola** [reply]
Еслиб можно было url без протокола записать, либо давайте закреплять https://  и нам нужен домен под нашим контролем

**Reply by niquola** [reply]
Чисто технически url выглядят так себе в поисках и данных. Как минимум я тогда за максимально короткие - `https://fhir.ru/id/ogrn`. И видимо only https, хотя есть еще PRURL которые с // начинаются - но скорее всего половина библиотек не поймет (хотя стоит проверить) 

Для сравнения:

* `urn:ru:fhir:id:ogrn` 
* `https://fhir.ru/id/ogrn`
*  `//fhir.ru/id/ogrn` 
* `urn:oid:1233.67.5654.78.1`

**Comment by MariaAnf** [comment]
### Голосование: Как мы будем кодировать Identifier.system
**Процедура голосования**: голосуем пальцем вверх за выбранный вариант (находится внизу поста <img width="31" alt="Screenshot 2022-01-28 at 18 15 30" src="https://user-images.githubusercontent.com/58726738/151572200-981fd0c2-1382-40db-8b15-70dc16008b46.png"> )


## Вариант 1: Использовать URN 

Пример:

* `urn:ru:fhir:id:ogrn` - ОГРН  организации 
* `urn:ru:fhir:id:snils` - СНИЛС пациента

**Reply by Vrodvi** [reply]
Поддерживаю вариант с URN, но там где есть утверждённые OID - использовать их, например: urn:oid:1.2.643.100.3 для СНИЛС

**Comment by MariaAnf** [comment]
### Голосование: Как мы будем кодировать Identifier.system
**Процедура голосования**: голосуем пальцем вверх за выбранный вариант (находится внизу поста <img width="31" alt="Screenshot 2022-01-28 at 18 15 30" src="https://user-images.githubusercontent.com/58726738/151572200-981fd0c2-1382-40db-8b15-70dc16008b46.png"> )


## Вариант 2: Использовать URL

Пример:

* `https://fhir.ru/id/ogrn`
*  `https://fhir.ru/id/snils`

**Comment by ai-fhir** [comment]
@niquola @MariaAnf вчера на большом созвоне у нас появилось три вопроса/пожелания:
1. Хочется зафиксировать регламент голосования (кто-то явно голосовал стрелочкой, кто-то пальцем вверх; ставим ли палец вниз? если да, то обязательно ли в этом случае писать почему не согласен и как учитывать этот голос, если человек не написал обоснования; что делаем, если участники голосования не пришли к согласию?)
2. @Vrodvi предложил третий пункт голосования - разрешить разные типы идентификаторов, но ограничить набор вариантов следующими - url, urn, oid. Это усложнит информационные системы, т.к. они должны будут поддерживать все три типа идентификаторов, но облегчит переход текущих систем. 
3. @s-shvyrev попросил подготовить презентацию/документ, который бы обосновал почему FHIR предпочитает url oid-ам.

**Comment by eugenekogan-spb** [comment]
Решение на http://fhir-ru.zendoc.me/meetings.2022-03-29 (принято как рабочее, без голосования) - использовать  URL.

@niquola, видимо, надо уточнить это решение: предложить конкретный вид URL для различных случаев использования.

**Reply by niquola** [reply]
В согласии с соглашением по названиями профайлов - предлагаю использовать  что-то типа:

`http://fhir.ru/<ig>/systems/<lowercase(common-human-name)>`


```
http://fhir.ru/core/systems/snils
http://fhir.ru/core/systems/ogrn
http://fhir.ru/frmo/systems/minzdrav-oid

```

**Reply by Vrodvi** [reply]
Важно не забыть про то, что некоторые системы идентификации уже имеют устоявшиеся идентификаторы в формате OID\UUID\URL\URN и их необходимо указывать хотя бы для возможности поиска (например, для СНИЛС - oid 1.2.643.100.3).

И если уже решено переиначивать существующие, устоявшиеся, идентификаторы, то желательно завести реестр всех использованных в наших профилях систем идентификации с возможностью поиска как по текущим устоявшимся идентификаторам, так и по новым, переопределённым.

Каждую систему идентификации следует описать в ресурсе NamingSystem, в следующем формате:
```
{
  "id": "core-namingsystem-snils",
  "resourceType": "NamingSystem",  
  "status": "active",    
  "kind": "identifier",
  "name": "Страховой номер индивидуального лицевого счёта, СНИЛС",
  "text": {
    "div": "<div xmlns='http://www.w3.org/1999/xhtml'>Identifier: Страховой номер индивидуального лицевого счёта, СНИЛС, может быть известен под следующими идентификаторами:<ol>\n    <li>OID: 1.2.643.100.3</li>\n    <li>[предпочтительно] URI: http://fhir.ru/core/systems/snils</li>\n  </ol>\n</div>",
    "status": "generated"
  },
  "uniqueId": [
    {
      "type": "oid",
      "value": "1.2.643.100.3"
    },
    {
      "type": "uri",
      "preferred": true,
      "value": "http://fhir.ru/core/systems/snils",
    }
  ]
}
```
Визуально такой реестр может быть представлен в виде таблицы как тут - http://hl7.org/fhir/identifier-registry.html (но желательно также иметь возможность поиска, как например тут - https://registry.fhir.org/results?query=%22urn%3Aoid%3A2.16.840.1.113883.6.77%22&latestFilter=true).

**Reply by eugenekogan-spb** [reply]
Создана Issue https://github.com/fhir-ru/core/issues/109 для создания списка систем идентификации.
